### PR TITLE
Bug 1526534 - Initail Windows commit

### PR DESCRIPTION
**Pull request by markcor**

### CODE REVIEW

File: modules/roles_profiles/manifests/profiles/files_system_managment.pp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::files_system_managment {
```
Comment (dividehex) [inline modules/roles_profiles/manifests/profiles/files_system_managment.pp]: This seems like it should belong under a windows specific module like 'win_filesystem'.  I would create 2 classes under that such as:
win_filesystem::disable_last_access
win_filesystem::disable_8dot3

File: modules/roles_profiles/manifests/profiles/nxlog.pp [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::nxlog {
```
Comment (dividehex) [inline modules/roles_profiles/manifests/profiles/nxlog.pp]: Move all of this to a nxlog module

File: modules/roles_profiles/manifests/profiles/nxlog.pp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    case $::operatingsystem {
+        'Windows': {
+            require  win_packages::nxlog
+
+            $programfilesx86 = $facts['programfilesx86']
+
+            file { "${programfilesx86}\\nxlog\\cert\\papertrail-bundle.pem":
```
Comment (dividehex) [inline modules/roles_profiles/manifests/profiles/nxlog.pp]: Is this file really required if we are sending logs to log aggregators via syslog?  If this file is really needed, it should at least be updated since it is 5 years old

File: modules/roles_profiles/manifests/profiles/openssh.pp [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::openssh {
```
Comment (dividehex) [inline modules/roles_profiles/manifests/profiles/openssh.pp]: move all of this to it's own windows specific module.  win_opensshd

File: modules/shared/facts.d/win_env_fact.ps1 [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Source Code Form is subject to the terms of the Mozilla Public
```
Comment (dividehex) [inline modules/shared/facts.d/win_env_fact.ps1]: I would argue that this "shared" module is very windows specific since it has powershell facts.  I would rename it win_shared or win_common

File: modules/shared/facts.d/win_env_fact.ps1 [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Creates useful facts based off of system environment variables
+
+$systemdrive     = $env:systemdrive
+$system32        = "$env:systemdrive\\windows\\system32"
```
Comment (dividehex) [inline modules/shared/facts.d/win_env_fact.ps1]: Instead of executing $env again just to get the system drive, why not use the puppet var to define $system32.  I would also add the suffix _path to any path you are defining so this would be $system32_path.

File: modules/shared/facts.d/win_env_fact.ps1 [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+$programdata     = $env:programdata.replace("\","\\")
+$programfiles    = $env:ProgramW6432.replace("\","\\")
+$programfilesx86 = "$systemdrive\\Program Files (x86)"
+
+# Environment variables
+write-host "systemdrive=$env:systemdrive"
```
Comment (dividehex) [inline modules/shared/facts.d/win_env_fact.ps1]: When writing custom facts, I would like the name to indicate so and also indicate if it is os specific.  For example, 'systemdrive' should be 'custom_win_systemdrive'

File: modules/shared/facts.d/win_env_fact.ps1 [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+write-host "system32=$system32"
+write-host "programdata=$programdata"
+write-host "programfiles=$programfiles"
+write-host "programfilesx86=$programfilesx86"
+
+
+# Facts built off of environment variables
+write-host "roninprogramdata=$programdata\\PuppetLabs\\ronin"
+write-host "roninsemaphoredir=$programdata\\PuppetLabs\\ronin\\semaphore"
+write-host "tempdir=$env:systemdrive\\Windows\\Temp"
```
Comment (dividehex) [inline modules/shared/facts.d/win_env_fact.ps1]: you've already looked up $env:systemdrive, just use the var $systemdrive to create tempdir.
eg. 
write-host "tempdir=${systemdrive}\\Windows\\Temp"

File: modules/shared/facts.d/win_location_fact.ps1 [hunk 9]
```diff
@@ -0,0 +1,22 @@
+$DhcpDomain = ((Get-ItemProperty 'HKLM:SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters').'DhcpDomain')
```
Comment (dividehex) [inline modules/shared/facts.d/win_location_fact.ps1]: It is fine to leave this here for now but I would really like to have this written in ruby so we can use it with other non-windows OS like linux and macOS.  That way we can keep the use of determining dc/region data common to all things.

File: modules/roles_profiles/manifests/profiles/openssh.pp [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    case $::operatingsystem {
+        'Windows': {
+        $programfiles     = $facts['programfiles']
+        $systemdrive      = $facts['systemdrive']
+        $programdata      = $facts['programdata']
+        $pwrshl_run_scrpt = lookup('pwrshl_run_scrpt')
+        $sshscrpt         = '"C:\Program Files\OpenSSH-Win64\install-sshd.ps1"'
+
+            defined_classes::pkg::win_zip_pkg { 'OpenSSH-Win64':
```
Comment (dividehex) [inline modules/roles_profiles/manifests/profiles/openssh.pp]: This should be a class under win_packages that is then included here.

File: modules/roles_profiles/manifests/profiles/nxlog.pp (hunk 3 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/nxlog.pp]: I think this file is needed on the cloud side of things.

File: modules/defined_classes/manifests/pkg/win_exe_pkg.pp [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+define defined_classes::pkg::win_exe_pkg ($pkg, $creates=$undef, $package=$title, $install_options_string=[]) {
+
+$pkgdir       = lookup('loc_pkg_dir')
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/pkg/win_exe_pkg.pp]: In this case this is a static directory set in hiera. If we pass this as a parameter it would be the same value every time. Alternatively I can set it up as a fact if that is better.

File: modules/roles_profiles/manifests/profiles/nxlog.pp (hunk 3 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/nxlog.pp]: It is needed for AWS: https://github.com/mozilla-releng/OpenCloudConfig/blob/4978a75a872847deadac1f72eec8ad3bdbcf15a4/userdata/Configuration/nxlog/win10.conf#L130

Just a note the nxlog pieces are going to continue to grow and become more complex as as we add  support for additional OSes and location type support.

File: modules/defined_classes/manifests/pkg/win_exe_pkg.pp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dividehex) [reply modules/defined_classes/manifests/pkg/win_exe_pkg.pp]: If this is a static dir that doesn't change, then hard code it.  Or pass it as a param and do the lookup in the profile.

File: modules/roles_profiles/manifests/profiles/nxlog.pp (hunk 3 above)
Comment (dividehex) [reply modules/roles_profiles/manifests/profiles/nxlog.pp]: I know that recently changed on the cloud side but I still think logs should be sent to aggregators and not directly to papertrail.  We need to fix the issue causing logs to be broken when sent to the log aggregators in the first place.

File: modules/defined_classes/manifests/pkg/win_exe_pkg.pp (hunk 11 above)
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/pkg/win_exe_pkg.pp]: It is a static dir that is called in each of the 3 define type package manifests. My preference is not to hard code the same path multiple times or pass the same value for a parameter multiple times. If for what ever reason this dir doesn't work on a different architecture or OS version, we should have one location to change, so it may stay consistent across all Windows platforms.

File: modules/roles_profiles/files/windows/authorized_keys [hunk 12]
```diff
@@ -0,0 +1,13 @@
+ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDFV81vjWsDxiAI8SUaVuxBi+zFuOseRFhg2t+F3N85J2XPOW5VNPPuRyWbD9u/W94iWDmZQT0BuyOg69ElE2WuFnlIF/n+SbviBSt9fc3IPntcalM1IIpoDUElQkKIkyaenBBthQP2yOzDWO7g4KFdIhr9sl4CMTYwv0Mr0Hit3KKXTRh929/VZ9Tb2yI+bL8bAniA5IcJ0dFuPN2cHWdpqqAo4Qf6T515xgsXKrt76fFeBzIT6qcHo62WD9f1v4YyZJJ6b7ylMThlf34b7Pebw3wG9qoBtzLmjH7gZ0XsgbQCI4cdzOhLmXZw1lFv6yqCF+5cmtx0VPgbCycwAG74stcjdh8ee/tYLMcTEBLIh/gIKuBgfBK5miSI7WFIOgyXqRCxin4is6eC1XQUcT9D01I+n7B0qwPPQFtW8gJvtI5aWiLNmPSujEC04DEzgH1RMEGs93lsUqrrK26ywAQDGiD/hfMyfMa8sumCtckGEzpVzg2dpWTBuQ78PkWQ4AAfp8uAWhEAm9GKV6mH+Ydi43od5vrGEuyGsiHOH10zRjCgwFonGKfKYWeITXcnHkogrHs2unO8zHZeMrrplQInhYoWkVR5uChjdU/DHDRWAUpNf4FME14tLjuyNyPzjvn9tE4BhHmAYtHGnkXbKWz4+sGkRm0q/0osjp4UWKdQ0w== Rob Thijssen <rthijssen@gmail.com>
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/files/windows/authorized_keys]: I am going to drop this from this pr and add it back in once we have the user module.

File: modules/roles_profiles/manifests/profiles/nxlog.pp (hunk 3 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/nxlog.pp]: I think that is outside of the scope of this. 

To the best of my understanding cloud instances always went directly to papertrail. The recent proposed change was to send hardware logs to papertrail directly. To change this we need to bring Rob into the conversation. Also note that the AWS instance report to TaskCluster papertrail. Which also something we may want to change in the future. 

For now I have moved it into a conditional clause based on location.

File: hiera.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+---
+version: 5
+
+hierarchy:
+  # testing
+  - name: "secrets"
+    path: "C:\\ronin\\secrets.yaml"
```
[the remark this answers is not in the document]
Comment (markcor) [reply hiera.yaml]: Dropping this until with have the data directory in the repo. I am using separate file for dev/testing.

File: modules/defined_classes/manifests/pkg/win_exe_pkg.pp [hunk 14]
```diff
@@ -0,0 +1,24 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/pkg/win_exe_pkg.pp]: Moved to win_packages module.

File: modules/defined_classes/manifests/pkg/win_msi_pkg.pp [hunk 15]
```diff
@@ -0,0 +1,21 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/pkg/win_msi_pkg.pp]: Moved to win_packages module.

File: modules/defined_classes/manifests/pkg/win_zip_pkg.pp [hunk 16]
```diff
@@ -0,0 +1,25 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/pkg/win_zip_pkg.pp]: Moved to win_packages module.

File: modules/shared/facts.d/win_env_fact.ps1 (hunk 6 above)
Comment (markcor) [reply modules/shared/facts.d/win_env_fact.ps1]: Done.

File: modules/shared/facts.d/win_env_fact.ps1 (hunk 7 above)
Comment (markcor) [reply modules/shared/facts.d/win_env_fact.ps1]: Facts and applicable hiera values are now prefixed with custom_win.

File: modules/shared/facts.d/win_location_fact.ps1 (hunk 9 above)
Comment (markcor) [reply modules/shared/facts.d/win_location_fact.ps1]: Because of time I am going to continue to use powershell, but ultimately these will be converted to ruby.

File: modules/roles_profiles/manifests/profiles/openssh.pp (hunk 10 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/openssh.pp]: Moved to its own module, and is now being called from the ssh profile.

File: modules/shared/facts.d/win_env_fact.ps1 (hunk 8 above)
Comment (markcor) [reply modules/shared/facts.d/win_env_fact.ps1]: Done.

File: modules/defined_classes/manifests/srv/disable_service.pp [hunk 17]
```diff
@@ -0,0 +1,11 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/defined_classes/manifests/srv/disable_service.pp]: Moved to win_disable_service module and is being called by disabled service profile.

File: modules/roles_profiles/files/windows/sshd_config [hunk 18]
```diff
@@ -0,0 +1,8 @@
+# Use default settings with the exception of the following options derived from:
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/files/windows/sshd_config]: Moved to win_openssh module.

File: modules/roles_profiles/manifests/profiles/admin_tools.pp [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::admin_tools {
+
+    case $::operatingsystem {
+        'Windows': {
+            include win_packages::jq
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/admin_tools.pp]: Now being called by common tools profile.

File: modules/roles_profiles/manifests/profiles/debug_tools.pp [hunk 20]
```diff
@@ -0,0 +1,18 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/debug_tools.pp]: Now being called by common tools profile.

File: modules/roles_profiles/manifests/profiles/debug_tools.pp [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::debug_tools {
+
+    case $::operatingsystem {
+        'Windows': {
+            include win_packages::process_debug
+            include win_packages::process_debug
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/debug_tools.pp]: No longer an issue.

File: modules/roles_profiles/manifests/profiles/disable_services.pp [hunk 22]
```diff
@@ -10,6 +10,11 @@
                 running => false,
             }
         }
+        'Windows': {
+        $disbaled_services = ['wsearch', 'VSS', 'puppet']
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/disable_services.pp]: Done.

File: modules/roles_profiles/manifests/profiles/disable_windows_defender.pp [hunk 23]
```diff
@@ -0,0 +1,31 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::disable_windows_defender {
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/disable_windows_defender.pp]: done

File: modules/roles_profiles/manifests/profiles/disable_windows_update.pp [hunk 24]
```diff
@@ -0,0 +1,36 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::disable_windows_update {
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/disable_windows_update.pp]: Done

File: modules/roles_profiles/manifests/profiles/nxlog.pp (hunk 2 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/nxlog.pp]: Moved to win_nxlog module.

File: modules/roles_profiles/manifests/profiles/openssh.pp (hunk 4 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/openssh.pp]: done

File: modules/roles_profiles/manifests/profiles/disable_system_restore.pp [hunk 25]
```diff
@@ -0,0 +1,24 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+class roles_profiles::profiles::disable_system_restore {
```
[the remark this answers is not in the document]
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/disable_system_restore.pp]: Done

File: modules/shared/facts.d/win_env_fact.ps1 (hunk 5 above)
Comment (markcor) [reply modules/shared/facts.d/win_env_fact.ps1]: Moved to win_shared.

File: modules/roles_profiles/manifests/profiles/files_system_managment.pp (hunk 1 above)
Comment (markcor) [reply modules/roles_profiles/manifests/profiles/files_system_managment.pp]: Done.
